### PR TITLE
sketch of hessian and vibrational analysis implementation

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
## Summary by CodeRabbit

* **New Features**
  * Analytical Hessian support: requestable at runtime, exposed via a public getter, returned in ASE-compatible shape.
  * Runtime overrides to enable/disable stress, Hessian, and uncertainty (force variance) during inference and MD runs.

* **Documentation**
  * New vibrational analysis tutorial notebook and usage guide demonstrating analytical vs numerical Hessians and configuration examples.

* **Tests**
  * New unit and integration tests validating Hessian computation and ASE Vibrations compatibility.

### CODE REVIEW

**Review by PythonFZ — CHANGES_REQUESTED** [review]

**Comment by TobiasDijkhuis** [comment]
Bumping because I am still very interested in this!

File: apax/config/model_config.py [hunk 1]
```diff
@@ -195,6 +197,7 @@ class BaseModelConfig(BaseModel, extra="forbid"):
     empirical_corrections: list[EmpiricalCorrection] = []
 
     calc_stress: bool = False
+    calc_hessian: bool = False
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/config/model_config.py:225]: I see your point andthe handling of property heads follow a similar structure to your suggestion. For forces and stress, there are no real configuration options so I think a binary flag is fine. Realistically, we will never implement a numerical stress tensor into the model itself. The only option I could think of here would be something like non-conservative forces, but that would require a larger redesign anyway.

File: apax/md/ase_calc.py [hunk 2]
```diff
@@ -164,11 +237,24 @@ def __init__(
...
+        if "stress" in self.implemented_properties:
+            model = ProcessStress().apply(model)
+
+        for transformation in self.transformations:
+            model = transformation.apply(model)
+
+        return model
+
     def _update_implemented_properties(self):
         """
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/md/ase_calc.py:265]: The model's capability is defined at model init but may be augmented via function transformations (e.g. UDD). Stress is handled differently during training and inference. THe ensemble transformation comes from the way model ensembles are implemented. Originally it was convenient to simply vmap the train step to get ensembles. With the shallow ensemble this leads to a divergence of how the two types of ensembles need to be treated in the code. ultimately, the full ensemble should probably be encapsulated into a dedicated model class.

File: apax/md/simulate.py [hunk 3]
```diff
@@ -540,8 +540,19 @@ def md_setup(model_config: Config, md_config: MDConfig):
         bias_list = [BiasEnergies(b.model_dump()) for b in md_config.biases]
         biases.extend(bias_list)
 
+    force_variance = "forces_uncertainty" in md_config.properties
+    if md_config.dynamics_checks:
+        for check in md_config.dynamics_checks:
+            if check.name == "forces_uncertainty":
+                force_variance = True
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/md/simulate.py:565]: The force variance option only exists for shallow ensemble forces. An SE always predicts energy uncertainty (there is no downside), force uncertainty is expensive to you may want to turn it off and for stress it is not implemented at all because we have not seen a purpose for that yet.

File: apax/nn/builder.py [hunk 4]
```diff
@@ -200,23 +203,34 @@ def build_energy_derivative_model(
...
+            calc_hessian = self.config["calc_hessian"]
+
         if (
             self.config["ensemble"]
             and self.config["ensemble"]["kind"] == "shallow"
             and self.config["ensemble"]["n_members"] > 1
         ):
+            if force_variance is None:
+                force_variance = self.config["ensemble"]["force_variance"]
+
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/nn/builder.py:219]: This can't happen because the outer conditional checks for the ensemble key.

File: tests/integration_tests/md/test_ase_hessian.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    builder = Builder(model_config.model.model_dump(), n_species=n_species)
+    model = builder.build_energy_derivative_model(apply_mask=False)
+
+    rng_key = jax.random.PRNGKey(model_config.seed)
+
+    # Minimal input for init
+    R = jnp.asarray(positions, dtype=jnp.float64)
+    Z = jnp.asarray(atomic_numbers)
+    idx = jnp.array([[1, 2, 0, 2, 0, 1], [0, 0, 1, 1, 2, 2]])
+    offsets = jnp.zeros((6, 3), dtype=jnp.float64)
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply tests/integration_tests/md/test_ase_hessian.py:42]: neighborlists and cell shifts. common names for these quantities in the code. We should probably make a dataclass for this.

File: tests/unit_tests/model/test_hessian.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+from apax.nn.models import EnergyDerivativeModel, EnergyModel
+
+
+def test_hessian_prediction():
+    # Use float64 for better precision in Hessian
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply tests/unit_tests/model/test_hessian.py:9]: mixed precision is the default. It's configurable.

**Comment by M-R-Schaefer** [comment]
@TobiasDijkhuis I have tested the Hessian training implementation on the hessian qm9 dataset you linked. Including it in the loss certainly improves the hessian metrics, but care needs to be taken so it doesn't interfere with the quality of forces. The dataset is fairly challenging because the forces are almost all 0 or have low magnitudes while the hessian elements can be huge. My current solution is to  use a mass weighted hessian loss which somewhat reduces the dynamic range and keeps most of the force prediction quality. You can see the label histograms of the subset I used below


label_distributions



There are more involved ways to implement hessian training, the most promising of which is probably to perform hessian vector product matching with vectors that favor the low frequency subspace, but this would require more time than I can currently offer.
Feel free to try it out and share your experience. The documentation features a recommended workflow.
